File: seminar_4/task_2/shatrov.igor/Task2.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;
                                                   //8 9 11 7 16 3 1
public class Task2 {
    public static void main(String[] args) {
        Scanner scn = new Scanner(System.in);

        List<Integer> arr = Arrays.stream(scn.nextLine().split(" "))
                .map(Integer::parseInt)
                .collect(Collectors.toList());

        MyBinaryTreeNode root = Task1.buildTree(arr, 0);

        System.out.println(isSymmetric(root));
    }

    public static boolean isSymmetric(MyBinaryTreeNode root) {
        if (root == null) {
            return true;
        }

        ArrayList<Integer> lst = new ArrayList<>();

        search(root, lst);

        for (int i = 0; i < lst.size(); i++) {
            if (!lst.get(i).equals(lst.get(lst.size() - i - 1))) {
                return false;
            }
        }

        return true;
    }

    public static void search(MyBinaryTreeNode node, ArrayList<Integer> lst) {
        if (node == null) {
            return;
        }

        search(node.left, lst);
        lst.add(node.val);
        search(node.right, lst);
    }
}

File: seminar_4/task_2/task_2.py
# Понять симметричное ли дерево 
# Задача 2

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_symmetric(root):
    if not root:
        return True
    return is_mirror(root.left, root.right)

def is_mirror(left, right):
    if not left and not right:
        return True
    if not left or not right:
        return False
    return (left.val == right.val) and is_mirror(left.left, right.right) and is_mirror(left.right, right.left)


